// mongoosejs-cli/lib/commands/migration_generate.js
'use strict';

var _yargs = require('../core/yargs');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _helpers = require('../helpers');

var _helpers2 = _interopRequireDefault(_helpers);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _cliColor = require('cli-color');

var _cliColor2 = _interopRequireDefault(_cliColor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.builder = yargs => (0, _yargs._baseOptions)(yargs).option('name', {
  describe: 'Defines the name of the migration',
  type: 'string',
  demandOption: true
}).argv;

const generateMigrationName = args => {
  return _lodash2.default.trimStart(_lodash2.default.kebabCase(args.name), '-');
};

exports.handler = function (args) {
  _helpers2.default.init.createMigrationsFolder();

  _fs2.default.writeFileSync(_helpers2.default.path.getMigrationPath(generateMigrationName(args)), _helpers2.default.template.render('migrations/skeleton.js', {}, {
    beautify: false
  }));

  _helpers2.default.view.log('New migration was created at', _cliColor2.default.blueBright(_helpers2.default.path.getMigrationPath(generateMigrationName(args))), '.');

  process.exit(0);
};
