// mongoosejs-cli/lib/helpers/config-helper.js
'use strict';

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _index = require('./index');

var _index2 = _interopRequireDefault(_index);

var _yargs = require('../core/yargs');

var _yargs2 = _interopRequireDefault(_yargs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const args = (0, _yargs2.default)().argv;

const api = {
  config: undefined,
  rawConfig: undefined,
  error: undefined,

  init() {
    return _bluebird2.default.resolve().then(() => {
      let config;

      if (args.url) {
        config = api.parseDbUrl(args.url);
      } else {
        try {
          config = require(api.getConfigFile());
        } catch (e) {
          api.error = e;
        }
      }
      return config;
    }).then(config => {
      if (typeof config === 'object' || config === undefined) {
        return config;
      } else if (config.length === 1) {
        return _bluebird2.default.promisify(config)();
      } else {
        return config();
      }
    }).then(config => {
      api.rawConfig = config;
    }).then(() => {
      // Always return the full config api
      return api;
    });
  },

  getConfigFile() {
    if (args.config) {
      return _path2.default.resolve(process.cwd(), args.config);
    }

    const defaultPath = _path2.default.resolve(process.cwd(), 'config', 'config.json');
    const alternativePath = defaultPath.replace('.json', '.js');

    return _index2.default.path.existsSync(alternativePath) ? alternativePath : defaultPath;
  },

  getModelsIndexFile() {
    return _index2.default.path.getModelPath('index');
  },

  relativeConfigFile() {
    return _path2.default.relative(process.cwd(), api.getConfigFile());
  },

  configFileExists() {
    return _index2.default.path.existsSync(api.getConfigFile());
  },

  getDefaultConfig() {
    return JSON.stringify({
      development: {
        database: {
          url: 'mongodb://localhost/mongoose_dev',
          options: {
            useNewUrlParser: true
          }
        }
      },
      test: {
        database: {
          url: 'mongodb://localhost/mongoose_test',
          options: {
            useNewUrlParser: true
          }
        }
      },
      production: {
        database: {
          protocol: 'mongodb',
          username: 'root',
          password: 'password',
          name: 'database_production',
          host: 'localhost',
          port: '',
          options: {
            useNewUrlParser: true
            //dbName: "" // uncomment this line if you use something like mongo atlas
          }
        }
      }
    }, undefined, 2) + '\n';
  },

  writeDefaultConfig() {
    const configPath = _path2.default.dirname(api.getConfigFile());

    if (!_index2.default.path.existsSync(configPath)) {
      _index2.default.asset.mkdirp(configPath);
    }

    _fs2.default.writeFileSync(api.getConfigFile(), api.getDefaultConfig());
  },

  readConfig() {
    if (!api.config) {
      const env = _index2.default.generic.getEnvironment();

      if (api.rawConfig === undefined) {
        throw new Error('Error reading "' + api.relativeConfigFile() + '". Error: ' + api.error);
      }

      if (typeof api.rawConfig !== 'object') {
        throw new Error('Config must be an object or a promise for an object: ' + api.relativeConfigFile());
      }

      if (args.url) {
        _index2.default.view.log('Parsed url ' + api.filteredUrl(args.url, api.rawConfig));
      } else {
        _index2.default.view.log('Loaded configuration file "' + api.relativeConfigFile() + '".');
      }

      if (api.rawConfig[env]) {
        _index2.default.view.log('Using environment "' + env + '".');

        api.rawConfig = api.rawConfig[env];
      }

      // The Sequelize library needs a function passed in to its logging option
      if (api.rawConfig.database.logging && !_lodash2.default.isFunction(api.rawConfig.database.logging)) {
        api.rawConfig.database.logging = console.log;
      }

      // in case url is present - we overwrite the configuration
      if (api.rawConfig.database.url) {
        api.rawConfig.database = _lodash2.default.merge(api.rawConfig.database, api.parseDbUrl(api.rawConfig.database.url));
      }

      api.config = api.rawConfig;
    }
    return api.config;
  },

  readConf() {
    try {
      api.config = require(api.getConfigFile());
      api.rawConfig = api.config;
    } catch (e) {
      throw new Error('Error occured when looking for "' + api.relativeConfigFile() + '". Kindly bootstrap the project using "mongoose init" comand.');
    }

    const env = _index2.default.generic.getEnvironment();

    if (api.rawConfig === undefined) {
      throw new Error('Error reading "' + api.relativeConfigFile() + '". Error: ' + api.error);
    }

    if (typeof api.rawConfig !== 'object') {
      throw new Error('Config must be an object: ' + api.relativeConfigFile());
    }

    _index2.default.view.log('Loaded configuration file "' + api.relativeConfigFile() + '".');

    if (api.rawConfig[env]) {
      _index2.default.view.log('Using environment "' + env + '".');

      api.rawConfig = api.rawConfig[env];
    }

    if (api.rawConfig.database.logging && !_lodash2.default.isFunction(api.rawConfig.database.logging)) {
      api.rawConfig.database.logging = console.log;
    }

    // in case url is present - we overwrite the configuration
    if (api.rawConfig.database.url) {
      api.rawConfig.database = _lodash2.default.merge(api.rawConfig.database, api.parseDbUrl(api.rawConfig.database.url));
    }

    return api.rawConfig;
  },

  filteredUrl(uri, config) {
    const regExp = new RegExp(':?' + (config.password || '') + '@');
    return uri.replace(regExp, ':*****@');
  },

  urlStringToConfigHash(urlString) {
    try {
      const urlParts = _url2.default.parse(urlString);
      let result = {
        name: urlParts.pathname.replace(/^\//, ''),
        host: urlParts.hostname,
        port: urlParts.port ? urlParts.port : '27017',
        protocol: urlParts.protocol.replace(/:$/, ''),
        ssl: urlParts.query ? urlParts.query.indexOf('ssl=true') >= 0 : false
      };

      if (urlParts.auth) {
        result = _lodash2.default.assign(result, {
          username: urlParts.auth.split(':')[0],
          password: urlParts.auth.split(':')[1]
        });
      }

      return result;
    } catch (e) {
      throw new Error('Error parsing url: ' + urlString);
    }
  },

  parseDbUrl(urlString) {
    return api.urlStringToConfigHash(urlString);
  }
};

module.exports = api;
