// mongoosejs-cli/lib/helpers/migration-helper.js
'use strict';

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _index = require('./index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const Mongoose = _index2.default.generic.getMongoose();

module.exports = {
  getCollectionName(modelName) {
    return Mongoose.pluralize(modelName);
  },

  generateCollectionCreationFileContent(args) {
    return _index2.default.template.render('migrations/create-table.js', {
      tableName: this.getCollectionName(args.name),
      attributes: _index2.default.model.transformAttributes(args.attributes),
      createdAt: 'created_at',
      updatedAt: 'updated_at'
    });
  },

  generateMigrationName(args) {
    return _lodash2.default.trimStart(_lodash2.default.kebabCase('create-' + args.name), '-');
  },

  generateCollectionCreationFile(args) {
    const migrationName = this.generateMigrationName(args);
    const migrationPath = _index2.default.path.getMigrationPath(migrationName);

    _index2.default.asset.write(migrationPath, this.generateCollectionCreationFileContent(args));
  }
};
