// mongoosejs-cli/lib/helpers/umzug-helper.js
'use strict';

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _index = require('./index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const storage = {
  migration: 'mongodb',
  seeder: 'none'
};
const storageCollectionName = {
  migration: 'mongoose_migrations_meta',
  seeder: 'mongoose_seeder_data'
};

const storageJsonName = {
  migration: 'mongoose-migrations.json',
  seeder: 'mongoose-seeders.json'
};

module.exports = {
  getStorageOption(property, fallback) {
    return _index2.default.config.readConfig()[property] || fallback;
  },

  getStorage(type) {
    return this.getStorageOption(type + 'Storage', storage[type]);
  },

  getStoragePath(type) {
    const fallbackPath = _path2.default.join(process.cwd(), storageJsonName[type]);

    return this.getStorageOption(type + 'StoragePath', fallbackPath);
  },

  getCollectionName(type) {
    return this.getStorageOption(type + 'StorageCollectionName', storageCollectionName[type]);
  },

  getStorageOptions(type, extraOptions) {
    const options = {};

    if (this.getStorage(type) === 'json') {
      options.path = this.getStoragePath(type);
    } else if (this.getStorage(type) === 'mongodb') {
      options.collectionName = this.getCollectionName(type);
    } else {
      options.collectionName = this.getCollectionName(type);
    }

    _lodash2.default.assign(options, extraOptions);

    return options;
  }
};
